packages/SwingSet/src/kernel/virtualObjectManager.js

```diff
@@ -315,6 +315,104 @@ export function makeVirtualObjectManager(
     });
   }
 
+  function vrefKey(value) {
+    const vobjID = valToSlot.get(value);
+    if (vobjID) {
+      const { type, virtual } = parseVatSlot(vobjID);
+      if (type === 'object' && virtual) {
+        return vobjID;
+      }
+    }
+    return undefined;
+  }
+
+  /* eslint max-classes-per-file: ["error", 2] */
+
+  const actualWeakMaps = new WeakMap();
+  const virtualObjectMaps = new WeakMap();
+
+  class VirtualizedWeakMap {
+    constructor() {
+      actualWeakMaps.set(this, new WeakMap());
+      virtualObjectMaps.set(this, new Map());
+    }
+
+    has(key) {
+      const vkey = vrefKey(key);
+      if (vkey) {
+        return virtualObjectMaps.get(this).has(vkey);
+      } else {
+        return actualWeakMaps.get(this).has(key);
+      }
+    }
+
+    get(key) {
+      const vkey = vrefKey(key);
+      if (vkey) {
+        return virtualObjectMaps.get(this).get(vkey);
+      } else {
+        return actualWeakMaps.get(this).get(key);
+      }
+    }
+
+    set(key, value) {
+      const vkey = vrefKey(key);
+      if (vkey) {
+        virtualObjectMaps.get(this).set(vkey, value);
+      } else {
+        actualWeakMaps.get(this).set(key, value);
+      }
+      return this;
+    }
+
+    delete(key) {
+      const vkey = vrefKey(key);
+      if (vkey) {
+        return virtualObjectMaps.get(this).delete(vkey);
+      } else {
+        return actualWeakMaps.get(this).delete(key);
+      }
+    }
+  }
+
```

Comment:
A nit. With the following, you should have a full emulation of the pre lockdown state. Once frozen you have an emulation of the post lockdown state, which will set `configurable` to `false.
```suggestion
Object.defineProperty(VirtualizedWeakMap, Symbol.toStringTag, { 
  value: 'WeakMap',
  writable: false,
  enumerable: false,
  configurable: true,
});
```
Same nit for WeakSet below.

Reply:
Yes, that's exactly the kind of weird JS voodoo I figured you'd know about.